src/systemd.c

```diff
@@ -0,0 +1,272 @@
+#include <assert.h>
+#include <errno.h>
+
+#define ONE_SEC 1000000
+
+static const char * const modes[] = {
+	"fail",			/* CGROUP_SYSTEMD_MODE_FAIL */
+	"replace",		/* CGROUP_SYSTEMD_MODE_REPLACE */
+	"isolate",		/* CGROUP_SYSTEMD_MODE_ISOLATE */
+	"ignore-dependencies",	/* CGROUP_SYSTEMD_MODE_IGNORE_DEPS */
+	"ignore-requirements",	/* CGROUP_SYSTEMD_MODE_IGNORE_REQS */
+};
+static_assert((sizeof(modes) / sizeof(modes[0])) == CGROUP_SYSTEMD_MODE_CNT,
+	      "modes[] array must be same length as CGROUP_SYSTEMD_MODE_CNT");
+
+static const char * const sender = "org.freedesktop.systemd1";
+static const char * const path = "/org/freedesktop/systemd1";
+static const char * const interface = "org.freedesktop.systemd1.Manager";
+
+int cgroup_set_default_scope_opts(struct cgroup_systemd_scope_opts * const opts)
+{
+	if (!opts)
+		return ECGINVAL;
+
+	opts->delegated = 1;
+	opts->mode = CGROUP_SYSTEMD_MODE_FAIL;
+	opts->pid = 0;
+
+	return 0;
+}
+
+/*
+ * Returns time elapsed in usec
+ *
+ * Inspired-by: https://github.com/cockpit-project/cockpit/blob/main/src/tls/socket-io.c#L39
+ */
+static int64_t elapsed_time(const struct timespec * const start, const struct timespec * const end)
+{
+	int64_t elapsed = (end->tv_sec - start->tv_sec) * 1000000 +
+			  (end->tv_nsec - start->tv_nsec) / 1000;
+
+	assert(elapsed >= 0);
+
+	return elapsed;
+}
+
+static int job_removed_callback(sd_bus_message *message, void *user_data, sd_bus_error *error)
+{
+	const char *result, *msg_path, *scope_name;
+	const char **job_path = user_data;
+	int ret;
+
+	ret = sd_bus_message_read(message, "uoss", NULL, &msg_path, &scope_name, &result);
+	if (ret < 0) {
+		cgroup_err("callback message read failed: %d\n", errno);
+		return 0;
+	}
+
+	if (*job_path == NULL || strcmp(msg_path, *job_path) != 0) {
+		cgroup_dbg("Received a systemd signal, but it was not our message\n");
+		return 0;
+	}
+
+	cgroup_dbg("Received JobRemoved signal for scope %s.  Result: %s\n", scope_name, result);
+
+	/*
+	 * Use the job_path pointer as a way to inform the original thread that the job has
+	 * completed.
+	 */
+	*job_path = NULL;
+	return 0;
+}
+
+int cgroup_create_scope(const char * const scope_name, const char * const slice_name,
+			const struct cgroup_systemd_scope_opts * const opts)
+{
+	sd_bus_message *msg = NULL, *reply = NULL;
+	sd_bus_error error = SD_BUS_ERROR_NULL;
+	const char *job_path = NULL;
+	struct timespec start, now;
+	sd_bus *bus = NULL;
+	pid_t child_pid;
+	int ret = 0;
+
+	if (!scope_name || !slice_name || !opts)
+		return ECGINVAL;
+
+	if (strcmp(&scope_name[strlen(scope_name) - strlen(".scope")], ".scope") != 0)
+		cgroup_warn("scope doesn't have expected suffix\n");
+	if (strcmp(&slice_name[strlen(slice_name) - strlen(".slice")], ".slice") != 0)
+		cgroup_warn("slice doesn't have expected suffix\n");
+
+	if (opts->mode >= CGROUP_SYSTEMD_MODE_CNT) {
+		cgroup_err("invalid systemd mode: %d\n", opts->mode);
+		return ECGINVAL;
+	}
+
+	if (opts->pid == 0) {
```

Comment:
Minor nitpick, should we initialize the `opts->pid = -1`. In the fork() control block, it makes the logic easier to read, where `pid == 0` has a special meaning.

Reply:
Good call.  Done.